source/Halibut.Tests/TcpClientCloseImmediatelyFixture.cs

```diff
@@ -0,0 +1,189 @@
+
+            using (disposables)
+            {
+                // Ensure the timeouts are not used
+                sut.WriteTimeout = (int)TimeSpan.FromSeconds(240).TotalMilliseconds;
+                sut.ReadTimeout = (int)TimeSpan.FromSeconds(240).TotalMilliseconds;
+
+                var data = new byte[655360];
+                var r = new Random();
+                r.NextBytes(data);
+
+                var semaphore = new SemaphoreSlim(1, 1);
+
+                // Brute force attempt to get the Write to be slow
+                Logger.Information("Start writing");
+                _ = Task.Run(async () =>
+                {
+                    await Try.RunTillExceptionOrCancellation(
+                        async () =>
+                        {
+                            semaphore.Release();
+                            await semaphore.WaitAsync(writeTokenSource.Token);
+                            Logger.Information("WRITER: Start");
+                            await sut.WriteToStream(StreamWriteMethod.WriteAsync, data, 0, data.Length, writeTokenSource.Token);
+                            Logger.Information("WRITER: Finished");
+                        },
+                        CancellationToken);
+                });
+
+                Logger.Information("Wait for writer to block...");
+                // Wait for some brute force writing...
+                while (semaphore.Wait(1000))
+                {
+                    // Do nothing, just wait for the semaphore to time out,
+                    // as that means writing is blocked
+                    Logger.Information("Keep waiting...");
+                    await Task.Delay(10);
+                }
+
+                Logger.Information("Writer has blocked, closing...");
+                var stopWatch = Stopwatch.StartNew();
+                client.CloseImmediately();
+                stopWatch.Stop();
+                Logger.Information($"Close completed, duration: {stopWatch.Elapsed.TotalMilliseconds}ms");
+
+                stopWatch.ElapsedMilliseconds.Should().BeLessThan(10);
+            }
+        }
+        
+        
+        [Test]
+        public async Task CanBeCalledMultipleTimes()
+        {
+            using var writeTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(120));
+            var (disposables, sut, client) = await BuildTcpClientAndTcpListener(
+                CancellationToken,
+                onListenerRead: async _ => await DelayForeverToTryAndDelayWriting(CancellationToken));
+
+            using (disposables)
+            {
+                // Ensure the timeouts are not used
+                sut.WriteTimeout = (int)TimeSpan.FromSeconds(240).TotalMilliseconds;
+                sut.ReadTimeout = (int)TimeSpan.FromSeconds(240).TotalMilliseconds;
+
+                var data = new byte[655360];
+                var r = new Random();
+                r.NextBytes(data);
+
+                var semaphore = new SemaphoreSlim(1, 1);
```

Comment:
nit: copy paste of all the logic to get the Socket into the correct state. Can this be refactored so it's shared across the 2 tests.